src/twisted/web/_cgi.py

```diff
@@ -0,0 +1,838 @@
+        last_boundary = next_boundary + b"--"
+        last_line_lfend = True
+        while True:
+            line = self.fp.readline(1 << 16)
+            self.bytes_read += len(line)
+            if not line:
+                self.done = -1
+                break
+            if line.endswith(b"--") and last_line_lfend:
+                strippedline = line.strip()
+                if strippedline == next_boundary:
+                    break
+                if strippedline == last_boundary:
+                    self.done = 1
+                    break
+            last_line_lfend = line.endswith(b"\n")
+
+    def make_file(self):
+        """Overridable: return a readable & writable file.
+
+        The file will be used as follows:
+        - data is written to it
+        - seek(0)
+        - data is read from it
+
+        The file is opened in binary mode for files, in text mode
+        for other fields
+
+        This version opens a temporary file for reading and writing,
+        and immediately deletes (unlinks) it.  The trick (on Unix!) is
+        that the file can still be used, but it can't be opened by
+        another process, and it will automatically be deleted when it
+        is closed or when the current process terminates.
+
+        If you want a more permanent file, you derive a class which
+        overrides this method.  If you want a visible temporary file
+        that is nevertheless automatically deleted when the script
+        terminates, try defining a __del__ method in a derived class
+        which unlinks the temporary files you have created.
+
+        """
+        if self._binary_file:
+            return tempfile.TemporaryFile("wb+")
+        else:
+            return tempfile.TemporaryFile("w+", encoding=self.encoding, newline="\n")
+
+
+def parse_multipart(fp, pdict, encoding="utf-8", errors="replace", separator="&"):
+    """Parse multipart input.
+
+    Arguments:
+    fp   : input file
+    pdict: dictionary containing other parameters of content-type header
+    encoding, errors: request encoding and error handler, passed to
+        FieldStorage
+
+    Returns a dictionary just like parse_qs(): keys are the field names, each
+    value is a list of values for that field. For non-file fields, the value
+    is a list of strings.
+    """
+    # RFC 2046, Section 5.1 : The "multipart" boundary delimiters are always
+    # represented as 7bit US-ASCII.
+    boundary = pdict["boundary"].decode("ascii")
+    ctype = "multipart/form-data; boundary={}".format(boundary)
+    headers = Message()
+    headers.set_type(ctype)
+    try:
+        headers["Content-Length"] = pdict["CONTENT-LENGTH"]
+    except KeyError:
+        pass
+    fs = FieldStorage(
+        fp,
+        headers=headers,
+        encoding=encoding,
+        errors=errors,
+        environ={"REQUEST_METHOD": "POST"},
+        separator=separator,
+    )
+    return {k: fs.getlist(k) for k in fs}
```

Comment:
I think that we should use the version of `cgi.parse_multipart` from before [this commit](https://github.com/python/cpython/commit/cc3fa204d357be5fafc10eb8c2a80fe0bca998f1). That commit regressed on the type of headers (from `bytes` to `str`) and introduced the encoding stuff, which is (a) a questionable design choice at this layer and (b) actively counterproductive for Twisted, since we need `bytes`.